cinema/views.py

```diff
@@ -21,24 +25,69 @@
 )
 
 
-class GenreViewSet(viewsets.ModelViewSet):
+# Anon: None
+# Is authenticated: list, retrieve
+# Is admin: create, update, partial_update, destroy
+class GenreViewSet(
+    mixins.CreateModelMixin,
+    mixins.ListModelMixin,
+    GenericViewSet,
+):
     queryset = Genre.objects.all()
     serializer_class = GenreSerializer
-
-
-class ActorViewSet(viewsets.ModelViewSet):
+    authentication_classes = (TokenAuthentication,)
+    permission_classes = (IsAdminOrIfAuthenticatedReadOnly,)
+
+    # def get_permissions(self):
+    #     if self.action in ("create", "update", "partial_update", "destroy"):
+    #         return [IsAdminUser()]
+    #
+    #     return super().get_permissions()  # It takes this
+    #     # value permission_classes = (IsAdminOrIfAuthenticatedReadOnly, )
+    #
+
+
+class ActorViewSet(
+    mixins.CreateModelMixin,
+    mixins.ListModelMixin,
+    GenericViewSet,
+):
     queryset = Actor.objects.all()
     serializer_class = ActorSerializer
+    authentication_classes = (TokenAuthentication,)
+    permission_classes = (IsAdminOrIfAuthenticatedReadOnly,)
 
 
-class CinemaHallViewSet(viewsets.ModelViewSet):
+class CinemaHallViewSet(
+    mixins.CreateModelMixin,
+    mixins.ListModelMixin,
+    GenericViewSet,
+):
     queryset = CinemaHall.objects.all()
     serializer_class = CinemaHallSerializer
+    authentication_classes = (TokenAuthentication,)
+    permission_classes = (IsAdminOrIfAuthenticatedReadOnly,)
 
 
-class MovieViewSet(viewsets.ModelViewSet):
+class MovieViewSet(
+    ReadOnlyModelViewSet,  # list and retrieve
+    mixins.CreateModelMixin,
+    GenericViewSet,
+):
     queryset = Movie.objects.prefetch_related("genres", "actors")
     serializer_class = MovieSerializer
+    authentication_classes = (TokenAuthentication,)
+    permission_classes = (IsAdminOrIfAuthenticatedReadOnly,)
+
+    def get_permissions(self):
+        # THIS METHOD IS USELESS since "retrieve" is allowed by default !!!
+        if self.action == "retrieve":
+            # These actions are available only for authenticated users
+            return [IsAuthenticated()]
+        # All other actions are available only for admins:
+        # ("update" , "partial-update", "delete")
+        return super().get_permissions()  # It takes this value:
```

Comment:
Yeah this code is redundant because IsAdminOrIfAuthenticatedReadOnly works the same